domain/application/service/application.go

```diff
@@ -227,26 +235,39 @@ type ApplicationService struct {
 	clock  clock.Clock
 
 	registry      storage.ProviderRegistry
-	secretService SecretService
+	secretDeleter SecretDeleter
 }
 
 // NewApplicationService returns a new service reference wrapping the input state.
-func NewApplicationService(st ApplicationState, params ApplicationServiceParams, logger logger.Logger) *ApplicationService {
+func NewApplicationService(
+	st ApplicationState, deleteSecretState DeleteSecretState,
+	params ApplicationServiceParams, logger logger.Logger,
+) *ApplicationService {
 	// Some uses of application service don't need to supply a storage registry,
 	// eg cleaner facade. In such cases it'd wasteful to create one as an
 	// environ instance would be needed.
 	if params.StorageRegistry == nil {
 		params.StorageRegistry = storage.NotImplementedProviderRegistry{}
 	}
-	if params.Secrets == nil {
-		params.Secrets = NotImplementedSecretService{}
+	if params.BackendAdminConfigGetter == nil {
+		params.BackendAdminConfigGetter = secretservice.NotImplementedBackendConfigGetter
+	}
+	if params.SecretBackendReferenceDeleter == nil {
+		params.SecretBackendReferenceDeleter = NotImplementedSecretDeleter{}
+	}
+	secretDeleter := &secretservice.SecretDeleter{
```

Comment:
This doesn't seem right. Why are we reliant on a service here? If it truly is a component, then the service can define an interface and operate on that. Then you could offer an implementation from the secret domain that provided that or have an application only version. Either way, we're coupling services here.

Reply:
There's complex logic in deleting a secret that is rightly implemented solely in the secret service. Services should be able to collaborate as part of a workflow. Either we allow services to be wired together via the dependency engine, we allow composition to be used, or we accept we need to cut and paste non trivial logic in different domains.